### DISCUSSION TITLE
Icons and CIDRAM loading time

### DISCUSSION THREAD

**Discussion by 737simpilot** [CIDRAM/CIDRAM#431] [General]
I was noticing some considerable slow loading of the frontend and upon checking it out in the browser Web Console it seems like it's all related to the icons. It says on average a 3 second load time for those icons actually. I don't exactly know why that would be the case since I see those files are not too large. Is there anything I can do or can be done in CIDRAM to mitigate that?

**Comment by Maikuolan** [comment]
Three different websites which have CIDRAM v3 installed at, which I'm maintaining at the moment. Two of them, the nav menu icons load within less than a few milliseconds (same as with my local dev copy), and due to using ETags plus some other browser caching mechanisms, should only even take that long to load the first time they're loaded, and for any subsequent page requests, be loaded instantly from the browser's own cache. The third website, for some reason, rather than the few milliseconds of the other two websites, takes a few hundred milliseconds to load each nav menu icon, and the ETags and various other caching mechanisms which CIDRAM utilises for static assets aren't being honoured at all, forcing a re-request of those static assets for every page load, and although I've been looking into it, at this point, I'm still not yet 100% as to why. If I'm able to figure out the cause, I should hopefully be able to come up with a decent solution.

I could probably just base64-encode all the nav menu icons, embed them directly into the CSS, and load them all that way, thus entirely avoiding the need for separate requests for each icon. That's what I've already doing for the pip icons and input decorations since a number of years back now. I'd been hoping to be able to have files be actual files though since introducing the nav menu icons, to be able to do things the more "proper" way, and also therefore allow easier customisation of themes and such (because directly replacing an image is probably easier for end-users than needing to re-encode an image and re-embed it into a custom CSS file), but I'll go the pip icons approach if it comes down to it that I'm not able to get to the bottom of why ETags and such aren't being properly honoured by what seems to be at least some websites.

Three whole seconds is definitely *way* too long. Even my experiences of the slowest, worst-performing website with CIDRAM v3 that I've seen, they still loaded in about ~400, ~500 milliseconds thereabouts (which is also still too long of a time IMHO). If it's taking three whole seconds, something is definitely going wrong beyond what I've experienced so far myself.

**Comment by 737simpilot** [comment]
It could very well be the fact I never keep cache and cookies on browser exit.

**Comment by Maikuolan** [comment]
Yeah.. That would probably do it. '^.^

Actually been thinking on the problem the past few hours, and thought.. I probably *could* embed everything into the CSS, as well as *also* keep the physical files available for people to customise if so desired: Just have some placeholders in the CSS, and have the files be called by CIDRAM and then their contents replace said placeholders.

Anyway, I'll see what I can do from my end.

**Comment by Maikuolan** [comment]
I pushed some changes just now which should help to improve the overall performance and load times regarding the icons. Let me know how it goes. :-)

**Comment by 737simpilot** [comment]
Definitely a huge improvement! Thanks! 👍
